### ISSUE TITLE
Incomplete/undocumented features regarding conversations API for proactive messaging

### ISSUE THREAD

**Issue by bandtank** [MicrosoftDocs/msteams-docs#8022]
This page does not give an example of how to send a request. A simple curl request would be more helpful than anything that is currently written. What is the endpoint to create a conversation? The basic pieces of information are missing.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 4ad9f167-7d29-f933-7b47-60bd744899b1
* Version Independent ID: 8097fcbd-adcc-ce49-a378-426c65af9296
* Content: [Send proactive messages - Teams](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/send-proactive-messages?tabs=dotnet)
* Content Source: [msteams-platform/bots/how-to/conversations/send-proactive-messages.md](https://github.com/MicrosoftDocs/msteams-docs/blob/main/msteams-platform/bots/how-to/conversations/send-proactive-messages.md)
* Service: **msteams**
* GitHub Login: @surbhigupta
* Microsoft Alias: **surbhigupta**

**Comment by Nivedipa-MSFT** [comment]
@bandtank - We will check and update it accordingly.

**Comment by bandtank** [comment]
Part of the problem is the lack of details about the endpoint for REST users. Creating a conversation for a channel or group conversation is not explained on the referenced page or anywhere else. None of the endpoints work the way the documentation suggests they should. It's useful to have code snippets, but it is also useful to see JSON objects being used for low level calls - that's where all the action happens and a few complete requests say more than several paragraphs of explanations.

Examples of requests that create a group with a few people and starting a new conversation in a channel would be extremely useful. I have been trying to do both of those things for days with no success at all.

**Comment by Nivedipa-MSFT** [comment]
@bandtank - We are checking this with engineering team. We will update you soon.

**Comment by bandtank** [comment]
For the purposes of documentation, I am posting more info here.

From [this](https://learn.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#conversationparameters-object) page, it says the following properties are required:

![image](https://user-images.githubusercontent.com/1170415/217059844-660a5eda-ced5-45e0-b9a9-700aa1cc457e.png)

Here is my attempt at creating a conversation and the associated error:

![image](https://user-images.githubusercontent.com/1170415/217059879-09afec58-650f-4bb2-87d1-ed0a8e56a98e.png)

None of the activity types make sense for this, so I don't know what to add. There are many other missing pieces of information as well. Here is the JSON data:

```
{
    // Is this required to create a conversation? If so, what goes here?
    "activity": "", 
    
    // Are other fields required?
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    
    // The documentation says "Channel-specific payload for creating the conversation."
    // What does that mean and where are the details?
    "channelData": "",
    
    // Requests from channel show this as true
    "isGroup": true,
    
    // Is this required to create a conversation in a channel?
    "members": "", 
    
    "tenantId": "{{tenantID}}",
    "topicName": "Test 123"
}
```

**Comment by Nivedipa-MSFT** [comment]
@bandtank - We are trying the API for channel and group chat. We will update you soon.

**Comment by bandtank** [comment]
I would like to clarify that the goal is to receive complete documentation. I managed to get the endpoint to work, but let me demonstrate what the real problem is.

This payload will create a conversation in a channel:
```
POST https://smba.trafficmanager.net/amer/v3/conversations
Body:
{
    "activity": {
        "type": "message",
        "text": "new conversation"
    },
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    "channelData":{
      "teamsChannelId":"{{teamID}}",
      "teamsTeamId":"{{teamID}}",
      "channel":{"id":"{{teamID}}"},
      "team":{"id":"{{teamID}}"},
      "tenant":{"id": "{{tenantID}}"}
    },
    "isGroup": true,
    "tenantId": "{{tenantID}}"
}
```
After searching this [link](https://github.com/microsoft/botframework-sdk/blob/main/specs/botframework-activity/botframework-activity.md#message-activity) for hours, I finally guessed it should be called a message. That makes sense in hindsight, but how are you supposed to know? The [documentation](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/channel-and-group-conversations?tabs=dotnet) that explains how to start a new conversation skips every important detail and the link to [proactive messaging](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/send-proactive-messages?tabs=dotnet#create-the-conversation) doesn't do any better. It's very likely that there are other important or even required fields, which I wouldn't know to send because nothing talks about the activity.

I finally got it to work by accident. After that happened, I started guessing at how to update the messages. Unfortunately, because there is no obvious way to get a list of messages from a conversation, I have to store every outbound message to every conversation thread. I finally realized it was possible to update a message after I guessed an endpoint and verb combination:

```
PUT https://smba.trafficmanager.net/amer/v3/conversations/{{conversationId}}/activities/{{activityId}}
{
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    "text": "# <strong>updating</strong> the first message",
    "textFormat": "markdown",
    "type": "message"
}
```

Finally, once I realized it was possible to update a message using the API. I found a [code example](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/update-and-delete-bot-messages?tabs=rest#update-messages) of the REST API requirements for this exact operation. Having that example for the creation of a conversation would have saved days of my time.

**Comment by Nivedipa-MSFT** [comment]
@bandtank - Could you please confirm should we update below content in following document or is there any content we need to add here?
https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/send-proactive-messages?tabs=dotnet



**Examples**
Be sure that you authenticate and have a [bearer token](https://learn.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-authentication#bot-to-connector) before creating a new conversation using the REST API.

**REST API to create a conversation in a one-one chat:**
HTTP
`POST {Service URL of your bot}/v3/conversations`

JSON
```
{
  "bot": {
    "id": "c38eda0f-e780-49ae-86f0-afb644203cf8",
    "name": "The Bot"
  },
  "members": [
    {
      "id": "29:012d20j1cjo20211"
    }
  ],
  "channelData": {
    "tenant": {
      "id": "197231joe-1209j01821-012kdjoj"
    }
  }
}

```
You can get the {Service URL of your bot} from TurnContext object like turnContext.Activity.ServiceURL parameter.

Provide id as your bot app ID and name as your bot name. You can get the members id from your bots TurnContext object such as turnContext.Activity.From.Id. Similarly, id of tenant, from your bots TurnContext object such as turnContext.Activity.ChannelData.Tenant.Id.

You must supply the user ID and the tenant ID. If the call succeeds, the API returns with the following response object.

JSON
```
{
  "id":"a:1qhNLqpUtmuI6U35gzjsJn7uRnCkW8NiZALHfN8AMxdbprS1uta2aT-jytfIlsZR3UZeg3TsIONNInBHsdjzj3PtfHuhkxxvS1jZZ61UAbw8fIdXcNSJyTJm7YvHFOgxo"
}
```
This ID is the personal chat's unique conversation ID. Store this value and reuse it for future interactions with the user.

**REST API to create a conversation in a channel:**

POST https://smba.trafficmanager.net/amer/v3/conversations
Body:
```

{
    "activity": {
        "type": "message",
        "text": "new conversation"
    },
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    "channelData":{
      "teamsChannelId":"{{teamID}}",
      "teamsTeamId":"{{teamID}}",
      "channel":{"id":"{{teamID}}"},
      "team":{"id":"{{teamID}}"},
      "tenant":{"id": "{{tenantID}}"}
    },
    "isGroup": true,
    "tenantId": "{{tenantID}}"
}
```
You can get the {Service URL of your bot} from TurnContext object like turnContext.Activity.ServiceURL parameter.
You can get the channelData from TurnContext object like turnContext.Activity.TeamsChannelData parameter.
Provide id as your bot app ID and name as your bot name. Similarly, id of tenant, from your bots TurnContext object such as turnContext.Activity.ChannelData.Tenant.Id.

**Update messages**
<html>
<body>
<p style="box-sizing: inherit; outline-color: inherit; margin: 1rem 0px 0px; padding: 0px; overflow-wrap: break-word; color: rgb(22, 22, 22); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">To update an existing activity within a conversation, include the<span> </span><code style="box-sizing: inherit; outline-color: inherit; font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; font-size: 13.6px; direction: ltr; background-color: var(--theme-inline-code); overflow-wrap: break-word; border-radius: 3px; padding: 0.1em 0.2em;">conversationId</code><span> </span>and<span> </span><code style="box-sizing: inherit; outline-color: inherit; font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; font-size: 13.6px; direction: ltr; background-color: var(--theme-inline-code); overflow-wrap: break-word; border-radius: 3px; padding: 0.1em 0.2em;">activityId</code><span> </span>in the request endpoint. To complete this scenario, you must cache the activity ID returned by the original post call.</p><div class="codeHeader" id="code-try-3" data-bi-name="code-header" style="box-sizing: content-box; outline-color: inherit; background-color: var(--theme-code-header); color: rgb(22, 22, 22); border-top-color: ; border-top-style: ; border-top-width: ; border-right-color: ; border-right-style: ; border-right-width: ; border-left-color: ; border-left-style: ; border-left-width: ; border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; min-height: 30px; border-bottom: 0px; flex-direction: row; margin-top: 16px; font-size: 0.8rem; display: flex; font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="language" style="box-sizing: inherit; outline-color: inherit; flex-grow: 1; padding: 2px 16px;">HTTP</span><button type="button" class="action position-relative display-none-print" data-bi-name="copy" style="box-sizing: inherit; outline-color: inherit; margin: 0px; font-family: inherit; font-size: 12.8px; line-height: normal; overflow: visible; text-transform: none; appearance: button; color: var(--theme-text); background-color: rgba(0, 0, 0, 0); cursor: pointer; position: relative !important; border-top-color: ; border-top-style: ; border-top-width: ; border-right-color: ; border-right-style: ; border-right-width: ; border-bottom-color: ; border-bottom-style: ; border-bottom-width: ; border-left-color: ; border-left-style: ; border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; border-left-width: 1px; align-items: center; padding: 2px 10px; display: flex;"><span class="icon margin-right-xxs" aria-hidden="true" style="box-sizing: inherit; outline-color: inherit; margin-inline-end: 0.5rem !important; justify-content: center; align-items: center; display: inline-flex; width: 1em; height: 1em;"><span class="docon docon-edit-copy" style="box-sizing: inherit; outline-color: inherit; font-family: docons; font-size: inherit; speak: none; font-variant: normal; text-transform: none; text-align: center; direction: ltr; -webkit-font-smoothing: antialiased; font-style: normal; font-weight: 400; line-height: 16px; display: inline-block;"></span></span><span style="box-sizing: inherit; outline-color: inherit;">Copy</span><div class="successful-copy-alert position-absolute right-0 top-0 left-0 bottom-0 display-flex align-items-center justify-content-center has-text-success-invert has-background-success is-transparent" aria-hidden="true" style="box-sizing: inherit; outline-color: var(--theme-success-invert); display: flex !important; justify-content: center !important; align-items: center !important; position: absolute !important; inset-block: 0px; inset-inline: 0px; color: var(--theme-success-invert) !important; background-color: var(--theme-success-base) !important; opacity: 0; transition: opacity 0.5s ease-in-out 0s;"><span class="icon font-size-lg" style="box-sizing: inherit; outline-color: inherit; font-size: 1.125rem !important; justify-content: center; align-items: center; display: inline-flex; width: 1em; height: 1em;"><span class="docon docon-check-mark" style="box-sizing: inherit; outline-color: inherit; font-family: docons; font-size: inherit; speak: none; font-variant: normal; text-transform: none; text-align: center; direction: ltr; -webkit-font-smoothing: antialiased; font-style: normal; font-weight: 400; line-height: 16px; display: inline-block;"></span></span></div></button></div><pre tabindex="0" class="has-inner-focus" style="box-sizing: inherit; outline-color: inherit; font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace !important; font-size: 0.875rem; margin: 0px; padding: 1rem; -webkit-font-smoothing: auto; border: 1px solid var(--theme-border); background-color: var(--theme-code-block); white-space: pre; word-spacing: 0px; word-break: normal; overflow-wrap: normal; tab-size: 4; hyphens: none; line-height: 1.3571; overflow: auto; color: rgb(22, 22, 22); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><code class="lang-http" data-author-content="PUT /v3/conversations/{conversationId}/activities/{activityId}
" style="box-sizing: inherit; outline-color: inherit; font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; font-size: 1em; direction: ltr; border: 0px; padding: 0px; line-height: 1.3571; display: block; position: relative;"><span style="box-sizing: inherit; outline-color: inherit;"><span class="hljs-attribute" style="box-sizing: inherit; outline-color: inherit; color: rgb(0, 104, 129);">PUT /v3/conversations/{conversationId}/activities/{activityId}
</span></span></code></pre><div class="has-inner-focus" style="box-sizing: inherit; outline-color: inherit; color: rgb(22, 22, 22); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Request | Response
-- | --
An [Activity ](https://learn.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#activity-object&preserve-view=true)object. | A [ResourceResponse ](https://learn.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#resourceresponse-object&preserve-view=true)object.

</div>
</body>
</html>To update an existing activity within a conversation, include the conversationId and activityId in the request endpoint. To complete this scenario, you must cache the activity ID returned by the original post call.

**Comment by bandtank** [comment]
That would be a great place to put updated documentation. I was also confused when I couldn't find any examples [here](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/conversation-messages?tabs=dotnet1%2Cdotnet2%2Ccsharp3%2Ccsharp4%2Ccsharp5%2Cdotnet#send-a-message). I understand there is a difference between proactive sending and replying, but not having any examples in a REST API tab made it difficult to understand what is required.

In short, the documentation appears to be fully explanative for the SDKs, but I would hope to see a REST API tab wherever there is example code. Even when you aren't using the REST API, seeing the low-level structures is often clarifying.

**Comment by Nivedipa-MSFT** [comment]
@bandtank - Should we add the content in below document or please redirect to the specific point where should we add it?
https://learn.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bot-conversations/bots-conv-proactive

**Comment by bandtank** [comment]
That is definitely one of the best places. Thanks for updating the documentation. I would also hope to see explanations about creating and updating (`POST` and `PUT`, respectively) [here](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/conversation-messages?tabs=dotnet1%2Cdotnet2%2Ccsharp3%2Ccsharp4%2Ccsharp5%2Cdotnet#send-a-message) in a new tab to the right of `JSON`, e.g. `REST API` or similar.

**Comment by Nivedipa-MSFT** [comment]
@bandtank - Thanks for your feedback. We are taking input from engineering team about updating of document.
We will let you know once we get any updates from them.

**Comment by Nivedipa-MSFT** [comment]
@surbhigupta12 - **Could you please add the content after following section of document.
https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/send-proactive-messages?tabs=typescript#proactively-install-your-app-using-graph**

**Contents need to be added:**

**Examples**
Be sure that you authenticate and have a [bearer token](https://learn.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-authentication#bot-to-connector) before creating a new conversation using the REST API.

**REST API to create a conversation in a one-one chat:**

```
`POST {Service URL of your bot}/v3/conversations`

JSON

{
   "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
  "members": [
    {
      "id": "{{memberID}}"
    }
  ],
  "channelData": {
    "tenant": {
      "id": "{{tenantID}}"
    }
  }
}

```

You can get the {Service URL of your bot} from TurnContext object like turnContext.Activity.ServiceURL parameter.
Provide id as your bot app ID and name as your bot name. 
You can get the members id from your bots TurnContext object such as turnContext.Activity.From.Id. 
Similarly, id of tenant, from your bots TurnContext object such as turnContext.Activity.ChannelData.Tenant.Id.

You must supply the user ID and the tenant ID. If the call succeeds, the API returns with the following response object.

JSON
```
{
  "id":"{{ personal chat's unique conversation ID}}"
}
```
This ID is the personal chat's unique conversation ID. Store this value and reuse it for future interactions with the user.

**REST API to create a conversation in a channel:**


```
POST {Service URL of your bot}/v3/conversations
JSON


{
    "activity": {
        "type": "message",
        "text": "new conversation"
    },
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    "channelData":{
      "teamsChannelId":"{{teamID}}",
      "teamsTeamId":"{{teamID}}",
      "channel":{"id":"{{teamID}}"},
      "team":{"id":"{{teamID}}"},
      "tenant":{"id": "{{tenantID}}"}
    },
    "isGroup": true,
    "tenantId": "{{tenantID}}"
}
```


You can get the {Service URL of your bot} from TurnContext object like turnContext.Activity.ServiceURL parameter.
You can get the channelData from TurnContext object like turnContext.Activity.TeamsChannelData parameter.
Provide id as your bot app ID and name as your bot name. Similarly, id of tenant, from your bots TurnContext object such as turnContext.Activity.ChannelData.Tenant.Id.
If the call succeeds, the API returns with the following response object.
```
{
    "id": "{{conversationID}}",
    "activityId": "{{activityID}}"
}
```

**REST API to Update message in conversation**
To update an existing activity within a conversation, include the conversationId and activityId in the request endpoint. To complete this scenario, you must cache the activity ID returned by the original post call.

```
PUT {Service URL of your bot}/v3/conversations/{conversationId}/activities/{activityId}
 
JSON
{
    "type": "message",
    "text": "This message has been updated"
}
```
To update an existing activity within a conversation, include the conversationId and activityId in the request endpoint. To complete this scenario, you must cache the activity ID returned by the original post call.
If the call succeeds, the API returns with the following response object.
```
{
    "id": "{{activityID}}"
}
```

Could you please update creating and updating (POST and PUT, respectively) [here](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/conversation-messages?tabs=dotnet1%2Cdotnet2%2Ccsharp3%2Ccsharp4%2Ccsharp5%2Cdotnet#send-a-message) in new tabs in the appropriate sections ("Send...", "Update...").

```
POST {Service URL of your bot}/v3/conversations`

JSON


{
   "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
  "members": [
    {
      "id": "{{memberID}}"
    }
  ],
  "channelData": {
    "tenant": {
      "id": "{{tenantID}}"
    }
  }
}


```

```
PUT {Service URL of your bot}/v3/conversations/{conversationId}/activities/{activityId}
 
JSON
{
    "type": "message",
    "text": "This message has been updated"
}

```
Add below content in following section
https://learn.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bot-conversations/bots-conv-proactive#creating-a-channel-conversation

Examples
```
POST {Service URL of your bot}/v3/conversations
JSON

{
    "activity": {
        "type": "message",
        "text": "new conversation"
    },
    "bot": {
        "id": "{{botID}}",
        "name": "{{botName}}"
    },
    "channelData":{
      "teamsChannelId":"{{teamID}}",
      "teamsTeamId":"{{teamID}}",
      "channel":{"id":"{{teamID}}"},
      "team":{"id":"{{teamID}}"},
      "tenant":{"id": "{{tenantID}}"}
    },
    "isGroup": true,
    "tenantId": "{{tenantID}}"
}

```
You can get the {Service URL of your bot} from TurnContext object like turnContext.Activity.ServiceURL parameter.
You can get the channelData from TurnContext object like turnContext.Activity.TeamsChannelData parameter.
Provide id as your bot app ID and name as your bot name. Similarly, id of tenant, from your bots TurnContext object such as turnContext.Activity.ChannelData.Tenant.Id.
If the call succeeds, the API returns with the following response object.
```
{
    "id": "{{conversationID}}",
    "activityId": "{{activityID}}"
}
```
